<a href="https://colab.research.google.com/github/BibikaShrestha/AI-Worksheets/blob/main/AIWorksheet5.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [6]:
from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [7]:
import pandas as pd
import numpy as np

In [8]:
df = pd.read_csv('/content/drive/MyDrive/student.csv')

To-do-1:
1. Read and Observe the Dataset.
2. Print top(5) and bottom(5) of the dataset {Hint: pd.head and pd.tail}.
3. Print the Information of Datasets. {Hint: pd.info}.
4. Gather the Descriptive info about the Dataset. {Hint: pd.describe}
5. Split your data into Feature (X) and Label (Y).

In [9]:

#1
print(df.head())

   Math  Reading  Writing
0    48       68       63
1    62       81       72
2    79       80       78
3    76       83       79
4    59       64       62


In [10]:
#2
print(df.tail())

     Math  Reading  Writing
995    72       74       70
996    73       86       90
997    89       87       94
998    83       82       78
999    66       66       72


In [11]:
#3
print(df.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 3 columns):
 #   Column   Non-Null Count  Dtype
---  ------   --------------  -----
 0   Math     1000 non-null   int64
 1   Reading  1000 non-null   int64
 2   Writing  1000 non-null   int64
dtypes: int64(3)
memory usage: 23.6 KB
None


In [23]:
#4
print(df.describe())

              Math      Reading      Writing
count  1000.000000  1000.000000  1000.000000
mean     67.290000    69.872000    68.616000
std      15.085008    14.657027    15.241287
min      13.000000    19.000000    14.000000
25%      58.000000    60.750000    58.000000
50%      68.000000    70.000000    69.500000
75%      78.000000    81.000000    79.000000
max     100.000000   100.000000   100.000000


In [12]:
missing_info= df.isnull().sum()/len(df)*100
print(missing_info)

Math       0.0
Reading    0.0
Writing    0.0
dtype: float64


In [24]:
for column in df.columns:
    if missing_info[column] > 10:
        df[column].fillna(df[column].median(), inplace=True)
    else:
        df.dropna(subset=[column], inplace=True)

In [25]:
#Checking if missing value is handled proeprly
df.isnull().sum()

Math       0
Reading    0
Writing    0
dtype: int64

In [17]:
#5
X=df[['Math','Reading']].values
y=df['Writing'].values

To - Do - 2:
1. To make the task easier - let’s assume there is no bias or intercept.
2. Create the following matrices:


In [27]:
from numpy.random import randn
n_samples, n_features = X_train.shape

W = np.random.randn(n_features, 1)

print("Initial random weights:")
print(W)
print("Shape of W:", W.shape)

Initial random weights:
[[1.1063291 ]
 [0.77163815]]
Shape of W: (2, 1)


To-do-3:
1. Split the dataset into training and test sets.
2. You can use an 80-20 or 70-30 split, with 80% (or 70%) of the data used for training and the rest for testing.



In [18]:

from random import shuffle
#making function for train_test split
def train_test_split_scratch(x,y,test_size=0.3,random_seed=42):
  np.random.seed(random_seed)
  indices=np.arange(X.shape[0])
  np.random.shuffle(indices) #Shuffle the indices
  test_split_size= int(len(X)*test_size)
  test_indices= indices[:test_split_size]
  train_indices= indices[test_split_size:]
  X_train=X[train_indices]
  y_train=y[train_indices]
  X_test=X[test_indices]
  y_test=y[test_indices]
  return X_train,X_test,y_train,y_test


In [21]:
#calling the train_test_split_scratch function
#Perform train-test Split
X_train, X_test, y_train, y_test = train_test_split_scratch(X, y, test_size=0.25)

In [22]:
#Output shapes to verify
print ("Shape of X_train:", X_train.shape)
print ("Shape of X_test:", X_test.shape)
print ("Shape of y_train:", y_train.shape)
print ("Shape of y_test:", y_test.shape)

Shape of X_train: (750, 2)
Shape of X_test: (250, 2)
Shape of y_train: (750,)
Shape of y_test: (250,)


To - Do - 4:

Feel free to build your own code or complete the given code:

In [30]:
def cost_function(X, Y, W):
    X = np.array(X, dtype=float)
    Y = np.array(Y, dtype=float).reshape(-1, 1)
    W = np.array(W, dtype=float).reshape(-1, 1)

    n = len(Y)
    Y_pred = X @ W
    error = Y_pred - Y
    cost = (1 / (2 * n)) * np.sum(error ** 2)

    return cost

To - Do - 5:

Make sure your code at To - Do - 4 passed the given test case:


In [31]:


X_test_case = np.array([[1, 2],
                        [3, 4],
                        [5, 6]])

Y_test_case = np.array([3, 7, 11])
W_test_case = np.array([[1], [1]])

cost = cost_function(X_test_case, Y_test_case, W_test_case)

if cost == 0:
    print("Proceed Further")
else:
    print("Something went wrong")

print("Cost:", cost)


Proceed Further
Cost: 0.0


To - Do - 6:

In [33]:

def gradient_descent(X, Y, W, alpha, iterations):
    Y = Y.reshape(-1, 1)
    m = len(Y)
    cost_history = []

    for _ in range(iterations):
        # Prediction
        Y_pred = X @ W

        # Error
        loss = Y_pred - Y

        # Gradient
        dw = (1 / m) * (X.T @ loss)

        # Update weights
        W = W - alpha * dw

        # Store cost
        cost = cost_function(X, Y, W)
        cost_history.append(cost)

    return W, cost_history


To - Do - 7:

In [34]:


np.random.seed(0)

X_gd = np.random.rand(100, 3)
Y_gd = np.random.rand(100)
W_gd = np.random.rand(3, 1)

alpha = 0.01
iterations = 1000

final_W, cost_history = gradient_descent(X_gd, Y_gd, W_gd, alpha, iterations)

print("Final Parameters:")
print(final_W)
print("Cost History (first 5 values):", cost_history[:5])


Final Parameters:
[[0.20551667]
 [0.54295081]
 [0.10388027]]
Cost History (first 5 values): [np.float64(0.10711197094660153), np.float64(0.10634880599939901), np.float64(0.10559826315680618), np.float64(0.10486012948320558), np.float64(0.1041341956428534)]


To - Do - 8:


In [35]:


def rmse(Y, Y_pred):
    Y = Y.reshape(-1, 1)
    return np.sqrt(np.mean((Y - Y_pred) ** 2))


To - Do - 9:

In [48]:
def r2(Y, Y_pred):
    Y = Y.reshape(-1, 1)
    mean_y = np.mean(Y)

    ss_tot = np.sum((Y - mean_y) ** 2)
    ss_res = np.sum((Y - Y_pred) ** 2)

    return 1 - (ss_res / ss_tot)


To - Do - 10:

We will define a function that:
1. Loads the data and splits it into training and test sets.
2. Prepares the feature matrix (X) and target vector (Y).
3. Defines the weight matrix (W) and initializes the learning rate and number of iterations.
4. Calls the gradient descent function to learn the parameters.
5. Evaluates the model using RMSE and R2

In [46]:
def main():
    # Load data
    df = pd.read_csv('/content/drive/MyDrive/student.csv')

    # Handle missing values (as you did)
    missing_info = df.isnull().sum() / len(df) * 100
    for column in df.columns:
        if missing_info[column] > 10:
            df[column].fillna(df[column].median(), inplace=True)
        else:
            df.dropna(subset=[column], inplace=True)

    # Features and target
    X = df[['Math', 'Reading']].values
    y = df['Writing'].values

    # Split
    X_train, X_test, y_train, y_test = train_test_split_scratch(X, y, test_size=0.25)

    # Initialize weights
    W_init = np.zeros((X_train.shape[1], 1))
    alpha = 0.00001
    iterations = 1000

    # Train
    W_final, cost_history = gradient_descent(X_train, y_train, W_init, alpha, iterations)

    # Predict
    y_pred_test = X_test @ W_final

    # Evaluate
    print("Final Weights:\n", W_final)
    print("RMSE:", rmse(y_test, y_pred_test))
    print("R² Score:", r2(y_test, y_pred_test))

    return W_final, cost_history

if __name__ == "__main__":
    main()

Iteration 0: Cost = 2014.1006
Iteration 100: Cost = 16.6507
Iteration 200: Cost = 16.1386
Iteration 300: Cost = 15.6682
Iteration 400: Cost = 15.2360
Iteration 500: Cost = 14.8390
Iteration 600: Cost = 14.4744
Iteration 700: Cost = 14.1394
Iteration 800: Cost = 13.8317
Iteration 900: Cost = 13.5490
Final Weights:
 [[0.3490432 ]
 [0.64534504]]
RMSE: 5.183049645617018
R² Score: 0.8897490516181759


To - Do - 11
Present your finding:
1. Did your Model Overfitt, Underfitts, or performance is acceptable.

Ans: The model shows acceptable performance—it’s not overfitting, nor is it underfitting. With an R² of 0.89 and an RMSE of around 5, it’s capturing a strong relationship between Math/Reading scores and Writing scores.

2. Experiment with different value of learning rate, making it higher and lower, observe the result.

Ans: The learning rate had a clear impact:

Too small (like 0.000001): slow but stable

Just right (around 0.00001): best results

Too high (0.0001 or more) : unstable, breaks learning